### PR TITLE
adding LAMBDA_INSIDE_VPC rule

### PR DESCRIPTION
**Pull request by rafihm**
I confirm these files are made available under CC0 1.0 Universal (https://creativecommons.org/publicdomain/zero/1.0/legalcode)

*Issue #, if available:*
#162 
*Description of changes:*
Adding a new rule LAMBDA_INSIDE_VPC

### CODE REVIEW

**Comment by jongogogo** [comment]
Please repush

**Comment by rafihm** [comment]
Automated build is showing pylint check result for two different rules (ELASTICSEARCH_IN_VPC_ONLY, EFS_ENCRYPTED_CHECK).

Local pylint test shows score 10/10 for this LAMBDA_INSIDE_VPC Rule.

File: python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    # if "vpcConfig" is not present, then lambda function is outside the VPC
+    if 'vpcConfig' not in configuration_item['configuration']:
+        return build_evaluation_from_config_item(configuration_item, 'NON_COMPLIANT', annotation='This Lambda Function is not in VPC')
```
Comment (jongogogo) [inline python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py]: Add a final dot to the annotation.

File: python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+    # if "vpcConfig" exists but no subnet is present, this scenario is possilbe if the lambda function was previously part of a VPC.
+    if not configuration_item['configuration']['vpcConfig']['subnetIds']:
+        return build_evaluation_from_config_item(configuration_item, 'NON_COMPLIANT', annotation='This Lambda Function is not in VPC')
```
Comment (jongogogo) [inline python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py]: Add a final dot to the annotation.

File: python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+
+    # compare the subnets of lambda with the input parameter using set operation
+    if set(configuration_item['configuration']['vpcConfig']['subnetIds']) - set(valid_rule_parameters):
+        return build_evaluation_from_config_item(configuration_item, 'NON_COMPLIANT', annotation='This Lambda Function is not associated with the subnets specified in the "subnetId" input parameter')
+
+    return build_evaluation_from_config_item(configuration_item, 'COMPLIANT')
+
+
+def evaluate_parameters(rule_parameters):
+    # if paramter is given, check whether its subnet Id or not, else ignore it as it is an optional paramter
```
Comment (jongogogo) [inline python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py]: typo in the comments

File: python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+    if "subnetId" not in rule_parameters:
+        return {}
+
+    # if paramter is given but value is empty then ignore it.
+    if not rule_parameters['subnetId']:
```
Comment (jongogogo) [inline python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py]: This cannot happen. Config will not pass an empty parameter.

File: python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        return {}
+
+    #split the paramter by delimter "," and remove whitespace, then validate the subnetId
```
Comment (jongogogo) [inline python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py]: typo

File: python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+    subnet_id_list = [each_subnet.strip() for each_subnet in rule_parameters['subnetId'].split(',')]
+    # to remove the emty item from the list (caused due trailing comma ',' in the parameter)
+    subnet_id_list = list(filter(None, subnet_id_list))
+
+    for each_subnet in subnet_id_list:
+        if "subnet-" not in each_subnet:
+            raise ValueError('Invalid value for paramter "subnetId", Expected subnet Id')
```
Comment (jongogogo) [inline python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py]: Typo
Please be more descriptive. The customer will read that :)

File: python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC_test.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+            return STS_CLIENT_MOCK
+        raise Exception("Attempting to create an unknown client")
+
+sys.modules['boto3'] = Boto3Mock()
+
+RULE = __import__('LAMBDA_INSIDE_VPC')
+
+class ComplianceTestScenarios(unittest.TestCase):
+
+    rule_emptry_parameter_value = '{"subnetId":""}'
```
Comment (jongogogo) [inline python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC_test.py]: typo in the name of the variable

File: python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC_test.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    def test_empty_parameter_value(self):
+        invoking_event = '{"configurationItem":{"configuration":' + json.dumps(self.lambda_inside_vpc) + ',"configurationItemCaptureTime":"2018-07-02T03:37:52.418Z","configurationItemStatus":"ResourceDiscovered","resourceType":"AWS::Lambda::Function","resourceId":"test_function"},"messageType":"ConfigurationItemChangeNotification"}'
+        response = RULE.lambda_handler(build_lambda_configurationchange_event(invoking_event, rule_parameters=self.rule_emptry_parameter_value), {})
+        resp_expected = []
+        resp_expected.append(build_expected_response('COMPLIANT', 'test_function', 'AWS::Lambda::Function'))
+        assert_successful_evaluation(self, response, resp_expected)
+
+    # Scenario 3.a
+    def test_lambda_outside_vpc(self):
+        invoking_event = '{"configurationItem":{"configuration":' + json.dumps(self.lambda_outside_vpc) + ',"configurationItemCaptureTime":"2018-07-02T03:37:52.418Z","configurationItemStatus":"ResourceDiscovered","resourceType":"AWS::Lambda::Function","resourceId":"test_function"},"messageType":"ConfigurationItemChangeNotification"}'
```
Comment (jongogogo) [inline python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC_test.py]: Can you create a function that does that. You do a lot of times the same thing in all scenarios :)

File: python/LAMBDA_INSIDE_VPC/parameters.json [hunk 9]
```diff
@@ -0,0 +1,12 @@
+{
+  "Version": "1.0", 
+  "Parameters": {
+    "CodeKey": "LAMBDA_INSIDE_VPC.zip", 
+    "SourceRuntime": "python3.6", 
+    "RuleName": "LAMBDA_INSIDE_VPC", 
+    "SourceEvents": "AWS::Lambda::Function", 
+    "OptionalParameters": "{}", 
+    "InputParameters": "{\"subnetId\": \"subnet-124231, subnet-3532432\"}"
```
Comment (jongogogo) [inline python/LAMBDA_INSIDE_VPC/parameters.json]: Move this into the OptionalParameters. And let the value empty for pushing into this repo. The RDK will do the right thing and not include them. If it is in InputParameter, the RDK will retrun an error if empty.

Comment (rafihm) [reply python/LAMBDA_INSIDE_VPC/parameters.json]: I tried giving below value for the `InputParameters` with the empty value for the subnetId and empty value to the InputParameters, 
```
    "InputParameters": "{\"subnetId\": \"\"}"
    "InputParameters": "{}"
```

In both cases RDK got deployed successfully and on the console it was showing Null as value for the SubnetId in the first and entire parameter as Null in second case. 

When i tried giving empty value on the console (GUI) it ignore if its empty and does not create any parameter at all. 

Looks like RDK is accepting the empty value, in that case should I keep the check or remove it ?

Comment (rafihm) [reply python/LAMBDA_INSIDE_VPC/parameters.json]: Sorry, i was referring the comment 

`This cannot happen. Config will not pass an empty parameter.`

File: python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py (hunk 4 above)
Comment (rafihm) [reply python/LAMBDA_INSIDE_VPC/LAMBDA_INSIDE_VPC.py]: I tried giving below value for the `InputParameters` with the empty value for the subnetId and empty value to the InputParameters,

```
    "InputParameters": "{\"subnetId\": \"\"}"
    "InputParameters": "{}"
```

In both cases RDK got deployed successfully and on the console it was showing Null as value for the SubnetId in the first and entire parameter as Null in second case.

When i tried giving empty value on the console (GUI) it ignore if its empty and does not create any parameter at all.

Looks like RDK is accepting the empty value, in that case should I keep the check or remove it ?

**Pull request merged**